src/components/Sessions/RequireSession.js

```diff
@@ -1,42 +1,124 @@
-import React from 'react';
+import React, {useContext, useEffect, useState} from 'react';
 import PropTypes from 'prop-types';
 import {Link} from 'react-router-dom';
-import {FormattedMessage} from 'react-intl';
+import {FormattedMessage, FormattedRelativeTime} from 'react-intl';
 
 import Anchor from 'components/Anchor';
 import Card from 'components/Card';
 import ErrorMessage from 'components/ErrorMessage';
+import Modal from "components/modals/Modal";
+import {Toolbar, ToolbarList} from "../Toolbar";
+import Button from "../Button";
+import {apiCall} from "../../api";
+import {ConfigContext} from "../../Context";
 
-// import SessionExpiry from './SessionExpiry';
+const WARN_SESSION_TIMEOUT_FACTOR = 0.9; // once 90% of the session expiry time has passed, show a warning
 
+const RelativeTimeToExpiry = ({numSeconds, raw = false}) => {
+  // more than 24 hours -> don't bother
+  if (numSeconds >= 3600 * 24) return null;
+  const component = (
+    <FormattedRelativeTime value={numSeconds} numeric="auto" updateIntervalInSeconds={1} />
+  );
+  if (raw) return component;
+  return <>&nbsp;({component})</>;
+};
+
+const untilWarningMilliseconds = (expiryDate) => {
+  /*
+  Calculate the time until the session expiry warning time should be displayed.
+   */
+  return (expiryDate - new Date()) * WARN_SESSION_TIMEOUT_FACTOR;
+};
+
+const RequireSession = ({expired = false, expiryDate = null, children}) => {
+  const {baseUrl} = useContext(ConfigContext);
+  const [secondsToExpiry, setSecondsToExpiry] = useState(0);
+  const [warningAccepted, setWarningAccepted] = useState(false);
+  const [warningVisible, setWarningVisible] = useState(false);
 
-const RequireSession = ({ expired=false, children }) => {
-  if (!expired) {
+  useEffect(() => {
+    /*
+    Show the warning if the session has expired.
+     */
+    if (expiryDate) {
+      const ms = untilWarningMilliseconds(expiryDate);
+      const timeout = setTimeout(() => {
+        setWarningVisible(true);
+      }, ms);
+      return () => clearTimeout(timeout);
+    }
+  }, [expiryDate]);
+
+  useEffect(() => {
+    /*
+    Count down to the session expiry time.
+     */
+    const interval = setInterval(() => {
+      setSecondsToExpiry(Math.floor((expiryDate - new Date()) / 1000));
+    }, 1000);
+    return () => clearInterval(interval);
+  }, [expiryDate]);
```

Comment:
This is quite a naive implementation because you're effectively doing a state update and component re-render every second, while we know that the warning only needs to be triggered in `90% of the remaining time`. With a session timeout of 15 minutes, that's 810 state updates that should have been only 1.